tuf_conformance/test_updater_delegation_graphs.py

```diff
@@ -0,0 +1,151 @@
+    for d in test_case.delegations:
+        if d.rolename in repo.mds:
+            targets = repo.mds[d.rolename].signed
+        else:
+            targets = Targets(1, spec_version, repo.safe_expiry, {}, None)
+        # unpack 'd' but skip "delegator"
+        role = DelegatedRole(*astuple(d)[1:])
+        repo.add_delegation(d.delegator, role, targets)
+
+    for target in test_case.target_files:
+        repo.add_artifact(*astuple(target))
+
+    if test_case.target_files:
+        repo.targets.version += 1
+    repo.update_snapshot()
+
+
+r"""
+Create a single repository with the following delegations:
+
+          targets
+*.doc, *md / \ release/*/*
+          A   B
+ release/x/* / \ release/y/*.zip
+            C   D
+
+Test that Updater successfully finds the target files metadata,
+traversing the delegations as expected.
+"""
+delegations_tree = DelegationsTestCase(
+    delegations=[
+        DelegationTester("targets", "A", paths=["*.doc", "*.md"]),
+        DelegationTester("targets", "B", paths=["releases/*/*"]),
+        DelegationTester("B", "C", paths=["releases/x/*"]),
+        DelegationTester("B", "D", paths=["releases/y/*.zip"]),
+    ],
+    target_files=[
+        TestTarget("targets", b"targetfile content", "targetfile"),
+        TestTarget("A", b"README by A", "README.md"),
+        TestTarget("C", b"x release by C", "releases/x/x_v1"),
+        TestTarget("D", b"y release by D", "releases/y/y_v1.zip"),
+        TestTarget("D", b"z release by D", "releases/z/z_v1.zip"),
+    ],
+)
+
+targets: DataSet = {
+    "no delegations": TargetTestCase("targetfile", True, []),
+    "targetpath matches wildcard": TargetTestCase("README.md", True, ["A"]),
+    "targetpath with separators x": TargetTestCase("releases/x/x_v1", True, ["B", "C"]),
+    "targetpath with separators y": TargetTestCase(
+        "releases/y/y_v1.zip", True, ["B", "D"]
+    ),
+    "targetpath is not delegated by all roles in the chain": TargetTestCase(
+        "releases/z/z_v1.zip", False, ["B"]
+    ),
+}
+
+
+targets_ids = targets.keys()
+targets_cases = targets.values()
+
+
+@pytest.mark.parametrize("target", targets_cases, ids=targets_ids)
+def test_targetfile_search(
+    client: ClientRunner, server: SimulatorServer, target: TargetTestCase
+) -> None:
+    exp_files = [*TOP_LEVEL_ROLE_NAMES, *target.visited_order]
+    exp_calls = [(role, 1) for role in target.visited_order]
+
+    init_data, repo = server.new_test(client.test_name)
+    assert client.init_client(init_data) == 0
+    init_repo(repo, delegations_tree)
+    exp_target = repo.artifacts[target.targetpath].target_file
+
+    # Call explicitly refresh to simplify the expected_calls list
+    client.refresh(init_data)
+    repo.metadata_statistics.clear()
+    client.download_target(init_data, target.targetpath)
+    if target.found:
+        assert client.download_target(init_data, target.targetpath) == 0
+        downloaded_target_path = client.get_downloaded_target_path()
```

Comment:
I still don't think we should do this. We decided in #106 that 

> tests never expect a specific filename to be written on disk

This is mostly because some clients may create a subdirectory structure for the targetpath, some might not (but also because of encoding issues in filenames). So we should not compare the filename to anything

Comparing the bytes should be enough, right?

Reply:
Added https://github.com/theupdateframework/tuf-conformance/pull/140#discussion_r1717372833 instead.